### PR TITLE
OTLP exporter refactor - pulling common code into BaseOtlpExporter class

### PR DESCRIPTION
**Pull request by alanwest**
Introduces new base class for `OtlpTraceExporter` and `OtlpMetricExporter`. Per the guidence here, I believe this is a non-breaking change: https://docs.microsoft.com/en-us/dotnet/core/compatibility/#types.

> ❓ REQUIRES JUDGMENT: Introducing a new base class
A type can be introduced into a hierarchy between two existing types if it doesn't introduce any new abstract members or change the semantics or behavior of existing types. For example, in .NET Framework 2.0, the DbConnection class became a new base class for SqlConnection, which had previously derived directly from Component.

### CODE REVIEW

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ResourceExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                if (oltpAttribute != null)
+                {
+                    processResource.Attributes.Add(oltpAttribute);
+                }
+            }
+
+            if (!processResource.Attributes.Any(kvp => kvp.Key == ResourceSemanticConventions.AttributeServiceName))
+            {
+                var serviceName = (string)ResourceBuilder.CreateDefault().Build().Attributes.Where(
+                    kvp => kvp.Key == ResourceSemanticConventions.AttributeServiceName).FirstOrDefault().Value;
```
Comment (alanwest) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ResourceExtensions.cs:43]: In moving this code, the logic here changed a little. When this code was on the `OtlpTraceExporter` class it was:

```csharp
var serviceName = (string)this.ParentProvider.GetDefaultResource().Attributes.Where(
                    kvp => kvp.Key == ResourceSemanticConventions.AttributeServiceName).FirstOrDefault().Value;
```

**Review by reyang — APPROVED** [review]
LGTM.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/BaseOtlpExporter.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+#else
+        internal Channel Channel { get; set; }
+#endif
+
+        internal OtlpExporterOptions Options { get; }
+
+        internal Metadata Headers { get; }
+
+        /// <inheritdoc/>
+        protected override bool OnShutdown(int timeoutMilliseconds)
```
Comment (reyang) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/BaseOtlpExporter.cs:65]: Might need to check the value to meet the semantic requirements https://github.com/open-telemetry/opentelemetry-dotnet/blob/6b7f2dd77cf9d37260a853fcc95f7b77e296065d/src/OpenTelemetry/BaseExporter.cs#L72.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/BaseOtlpExporter.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        {
+            if (this.Channel == null)
+            {
+                return true;
+            }
+
+            return Task.WaitAny(new Task[] { this.Channel.ShutdownAsync(), Task.Delay(timeoutMilliseconds) }) == 0;
```
Comment (reyang) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/BaseOtlpExporter.cs:72]: Do we want to support `-1` as Timeout.Infinite?

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/BaseOtlpExporter.cs (hunk 2 above)
Comment (alanwest) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/BaseOtlpExporter.cs:65]: `OnShutdown` is invoked from `BaseExporter.Shutdown` you've linked to, so the check is performed there. Are you suggesting we should add an additional check here?

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/BaseOtlpExporter.cs (hunk 3 above)
Comment (alanwest) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/BaseOtlpExporter.cs:72]: We support it today, but is your thought that maybe we shouldn't rely on an infinite timeout for shutting down the channel?

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ResourceExtensions.cs (hunk 1 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ResourceExtensions.cs:43]: I think we could simplify the linq a bit (combine Where + FirstOrDefault) as: `(string)this.ParentProvider.GetDefaultResource().Attributes.FirstOrDefault(kvp => kvp.Key == ResourceSemanticConventions.AttributeServiceName).Value`

**Review by CodeBlanch — APPROVED** [review]
LGTM

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/BaseOtlpExporter.cs (hunk 2 above)
Comment (reyang) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/BaseOtlpExporter.cs:65]: `BaseExporter.Shutdown` could take `-1` and pass it to `OnShutdown`, do we want to handle `-1` value as `Timeout.Infinite` in the `BaseOtlpExporter`?

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/BaseOtlpExporter.cs (hunk 3 above)
Comment (reyang) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/BaseOtlpExporter.cs:72]: Got it, I wasn't aware that `Task.Delay` already supports `-1`.
Just curious, if we do a special case would it give any benefit/improvement?

```csharp
if (timeoutMilliseconds == -1)
{
    return Task.Wait(this.Channel.ShutdownAsync());
}
else // here timeoutMilliseconds must be greater than or equal to 0
{
    return Task.WaitAny(new Task[] { this.Channel.ShutdownAsync(), Task.Delay(timeoutMilliseconds) }) == 0;
}
```

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/BaseOtlpExporter.cs (hunk 2 above)
Comment (reyang) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/BaseOtlpExporter.cs:65]: It's already addressed [here](https://github.com/open-telemetry/opentelemetry-dotnet/pull/2220#discussion_r681188810), just my lack of knowledge that `Task.Delay` already supports `-1`. 😅

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/BaseOtlpExporter.cs (hunk 3 above)
Comment (alanwest) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/BaseOtlpExporter.cs]: I don't think it hurts, so I made this change. Pretty sure it saves some allocations by not calling Task.Delay.

You got me looking at all the shutdown logic more closely, and I've realized that shutdown is currently only wired through from the TraceProvider and not the MetricProvider yet.

I noted the spec still had [shutdown as a TODO](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#shutdown). Do you think it makes sense to wait on the spec before wiring this up?

**Pull request merged**